Bluetooth: host: l2cap: dynamic channel bugfixes

jori-nordic opened the pull request:
See commit messages.

jori-nordic committed:
tests: Bluetooth: refactor l2cap stress test

- move all the context for a channel in its own custom struct
- allow -EAGAIN when sending, try sending again later

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: host: l2cap: trigger SDUs that get lost in limbo

Turns out the [first bugfix] was too naive: there is a case where resuming
all channels will not work on all queued SDUs, and the work handler will
give up and wait for the next sent SDU instead of trying to resume again.

This happens when the number of credits and conn contexts is very low for
the amount of data to send.

Always reschedule with a delay to avoid that situation.

[first bugfix]: https://github.com/zephyrproject-rtos/zephyr/pull/50476

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
tests: Bluetooth: make l2cap stress test fail

Running with a higher SDU length makes the devices get stuck at the start
of the transfer. To be fixed by next commits.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

hermabe commented on subsys/bluetooth/host/Kconfig.l2cap:
```diff
@@ -44,6 +44,15 @@ config BT_L2CAP_TX_MTU
 	  Maximum L2CAP MTU for L2CAP TX buffers.
 
+config BT_L2CAP_RESCHED_MS
+	int
```
Should this have a prompt so that it is user-configurable?

alwa-nordic commented:
Regarding commit https://github.com/zephyrproject-rtos/zephyr/pull/55519/commits/766b4bb8e1ec1ac4dbe71fd40ed0259a2f6f4c58:

Can you add a link to "first bugfix" in the commit message? (I assume it refers to an already merged PR.) I suggest using markdown syntax: "Turns out the \[first bugfix] was too naive" and "\[first bugfix]: https://github.com/zephyrproject-rtos/zephyr/pull/...".

alwa-nordic commented on subsys/bluetooth/host/l2cap.c:
I think it's bad to automatically give credits. Maybe the application is going to give more credits async (`bt_l2cap_chan_recv_complete`)? We don't know that the application is malfunctioning here. And yet, this code seems to assume that the application is malfunctioning.

jori-nordic replied:
that's not what's happening here. It's an API limitation: we only give complete SDUs to the application via a callback. The app doesn't know that it can call this `bt_l2cap_chan_recv_complete`.

jori-nordic replied:
I'm okay with aborting here btw. At least now, I've documented the initial credits value calculation.

alwa-nordic commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -147,7 +147,12 @@ struct bt_l2cap_le_endpoint {
-	/** Endpoint initial credits */
+	/** Endpoint initial credits
+	 *
+	 * The user should allocate enough credits to be able to transfer a full
+	 * SDU. 1 credit = 1 PDU of MPS length. Example:
+	 * SDU = 2000 bytes, MPS = 77 -> init_credits >= 26.
```
This is not enough to ensure successful transport. The remote may choose to underutilize MPS. The comment we are removing below hinted at exactly this.

https://github.com/zephyrproject-rtos/zephyr/blob/bb6fa5707f77d72307008405f19dd83247405392/subsys/bluetooth/host/l2cap.c#L2460-L2463

We need a contract with the application:

When the channel connected event handlers complete, `bt_l2cap_chan_ops.alloc_seg` shall be ready give out `init_credits` net_bufs. When the net_buf is destructor returns, a net_buf must be available from `alloc_seg` again. 

`bt_l2cap_chan_ops.alloc_buf` shall be ready give out one buffer after..
 - channel connected event handlers complete
 - when the channel's `recv` returns other than `-EINPROGRESS`
 - when the channel's `bt_l2cap_chan_recv_complete` is invoked

Contract is valid until the 'disconnected' event.

Additionally: (deprecated) If `bt_l2cap_chan_ops.alloc_seg` is NULL, then `alloc_buf` must in addition give out the buffers demanded of `bt_l2cap_chan_ops.alloc_seg`.

We make sure the data in a `seg` is copied to the `buf`, and `seg` is destroyed before we accept more ACL data from that conn.

alwa-nordic replied:
I realize the `-EINPROGRESS` situation is not completely fleshed out. I suppose we can demand `init_credits` buffers of `alloc_buf` if `-EINPROGRESS` is used. Seems wasteful though.

alwa-nordic replied:
When `-EINPROGRESS`, the `seg` start running out at the same rate as credits. This is exactly what we want. When `bt_l2cap_chan_recv_complete` is invoked, we can immediately grab a new `alloc_buf`, copy and destroy all received `segs`, and send back more credits.

jori-nordic replied:
@alwa-nordic I think you misunderstood how `.recv` works. It is only ever called after receiving a complete SDU.

This is the whole point of the PR, where the user basically has no levers it can pull to send more credits _when receiving PDUs_. It can only withhold credits by returning `-EINPROGRESS` after having received a full SDU.

Because this is the API that we have, IMO the credits should be an implementation detail, and should be managed by the host and not by the users at all. 

Even initial_credits is not a good idea with this API; I think the application should rather inform the stack of its SDU buffer size instead.

alwa-nordic commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -147,7 +147,12 @@ struct bt_l2cap_le_endpoint {
-	/** Endpoint initial credits */
+	/** Endpoint initial credits
+	 *
+	 * The user should allocate enough credits to be able to transfer a full
+	 * SDU. 1 credit = 1 PDU of MPS length. Example:
+	 * SDU = 2000 bytes, MPS = 77 -> init_credits >= 26.
```
This is not enough to ensure successful transport. The remote may choose to underutilize MPS. The comment we are removing below hinted at exactly this.

https://github.com/zephyrproject-rtos/zephyr/blob/bb6fa5707f77d72307008405f19dd83247405392/subsys/bluetooth/host/l2cap.c#L2460-L2463

We need a contract with the application:

When the channel connected event handlers complete, `bt_l2cap_chan_ops.alloc_seg` shall be ready give out `init_credits` net_bufs. When the net_buf is destructor returns, a net_buf must be available from `alloc_seg` again. 

`bt_l2cap_chan_ops.alloc_buf` shall be ready give out one buffer after..
 - channel connected event handlers complete
 - when the channel's `recv` returns other than `-EINPROGRESS`
 - when the channel's `bt_l2cap_chan_recv_complete` is invoked

Contract is valid until the 'disconnected' event.

Additionally: (deprecated) If `bt_l2cap_chan_ops.alloc_seg` is NULL, then `alloc_buf` must in addition give out the buffers demanded of `bt_l2cap_chan_ops.alloc_seg`.

We make sure the data in a `seg` is copied to the `buf`, and `seg` is destroyed before we accept more ACL data from that conn.

jori-nordic committed:
tests: Bluetooth: add l2cap MPS test

This test verifies the l2cap MPS workaround: when the other
side (non-zephyr host) sends PDUs that are < MPS.

The DUT is a host, while the Tester is a thin layer over the controller,
allowing fine control over l2cap packet contents.

Note that this test fails as-is. It is fixed by the next commits in the PR.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55519#issuecomment-1486448028 after 10:42:36.

codecoup-tester commented:
AutoPTS Bot results:
Failed tests:
L2CAP   L2CAP/COS/ECFC/BV-04-C          INCONC
L2CAP   L2CAP/ECFC/BV-43-C              FAIL - ERRATA TSE18351
L2CAP   L2CAP/ECFC/BV-44-C              INCONC - ERRATA TSE18351
L2CAP   L2CAP/TIM/BV-03-C               INCONC

jori-nordic commented:
Added DNM as I'm working on a test for the case when a peer uses less than MPS.

jori-nordic commented:
@sjanc could you re-run the PTS tests? thanks!

sjanc commented:
#AutoPTS run zephyr L2CAP

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55519#issuecomment-1494165302 after 13:43:06.

codecoup-tester commented:
AutoPTS Bot results:
Failed tests:
L2CAP   L2CAP/ECFC/BV-43-C              FAIL - ERRATA TSE18351
L2CAP   L2CAP/ECFC/BV-44-C              INCONC - ERRATA TSE18351
L2CAP   L2CAP/TIM/BV-03-C               INCONC

jori-nordic committed:
Bluetooth: host: l2cap: send only 1 credit

The API doesn't allow the stack to make any guarantees about the number of
available buffers that the app has.

Only send 1 credit at a time, as that is the only guarantee the stack can
give to the peer.

We can send MTU/MPS's worth of credits once we have acquired an SDU buffer
from the application (that is, on the first PDU of the SDU). Though we
still have to cap that to the buffer size we have just acquired.

------

The testcase added here shows a scenario where the relationship
between the number of credits and the number of available buffers does not
hold true any more:

In this test, the app only has one buffer in its pool.

The central will queue SDUs that are bigger than the stack's
buffers (so the user allocator is necessary) but lower than the
channel's MTU.

The device receiving the SDU keeps a reference to the buffer before
returning from the `recv` callback. It releases that reference after a
small delay.

The central will still have credits, so it will queue another SDU, but the
peripheral will not be able to receive the next SDU (as the allocator will
fail) and will close the channel.

To see the test fail, just revert the `l2cap.{c,h}` changes.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -861,25 +862,7 @@ static void l2cap_chan_rx_init(struct bt_l2cap_le_chan *chan)
-	    chan->rx.mtu + BT_L2CAP_SDU_HDR_SIZE) {
-		LOG_WRN("Not enough credits for a full packet");
-	}
+	atomic_set(&chan->rx.credits,  1);
```
minor: there should be only one space between `,` and `1` (now you have two)

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1653,7 +1628,7 @@ static void le_ecred_conn_rsp(struct bt_l2cap *l2cap, uint8_t ident,
 			/* Give credits */
 			l2cap_chan_tx_give_credits(chan, credits);
-			l2cap_chan_rx_give_credits(chan, chan->rx.init_credits);
+			atomic_set(&chan->rx.credits,  1);
```
same here

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1729,7 +1704,7 @@ static void le_conn_rsp(struct bt_l2cap *l2cap, uint8_t ident,
 		/* Give credits */
 		l2cap_chan_tx_give_credits(chan, credits);
-		l2cap_chan_rx_give_credits(chan, chan->rx.init_credits);
+		atomic_set(&chan->rx.credits,  1);
```
and here

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2303,38 +2249,18 @@ static void l2cap_chan_send_credits(struct bt_l2cap_le_chan *chan,
 
-	l2cap_chan_rx_give_credits(chan, credits);
+	__ASSERT_NO_MSG(atomic_get(&chan->rx.credits) == 0);
+	atomic_set(&chan->rx.credits,  1);
```
and here (seems like a copy-paste thing)

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -428,7 +428,15 @@ int bt_conn_send_cb(struct bt_conn *conn, struct net_buf *buf,
+			if (!(net_buf_pool_get(buf->pool_id) == &acl_tx_pool)) {
+#else
+			if (false) {
+#endif
```
Seems like the above comment was left in my cache (as you can see from the timestamp). Not sure if it's relevant or correct anymore, since the PR has probably evolved since then.

jori-nordic commented:
DNM again as throughput is 10x worse in HW tests. Will come up with something else

jori-nordic committed:
Bluetooth: host: att: clear ATT_PENDING_SENT

The ATT_PENDING_SENT flag was still not being cleared in all cases.

Also reset `data->att_chan` when not able to send on a given channel.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -539,7 +540,7 @@ static int l2cap_ecred_conn_req(struct bt_l2cap_chan **chan, int channels)
 	req->psm = sys_cpu_to_le16(ch->psm);
 	req->mtu = sys_cpu_to_le16(ch->rx.mtu);
 	req->mps = sys_cpu_to_le16(ch->rx.mps);
-	req->credits = sys_cpu_to_le16(ch->rx.init_credits);
```
@alwa-nordic What do we do with `init_credits` now? I suggest we keep it and just mark it as internal-use in l2cap.h. 
This way we don't break the build of existing apps.

alwa-nordic replied:
I agree.

jori-nordic commented:
@jhedberg could you re-review?

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2343,6 +2271,8 @@ int bt_l2cap_chan_recv_complete(struct bt_l2cap_chan *chan, struct net_buf *buf)
 	__ASSERT_NO_MSG(chan);
 	__ASSERT_NO_MSG(buf);
 
+	net_buf_unref(buf);
```
Is this correct? On line 2287 you're passing the buffer to `l2cap_chan_send_credits()` so you must hold a valid reference to the buffer until then.

jori-nordic replied:
yes. `buf` is overwritten in that fn. It's passed to `l2cap_create_le_sig_pdu` but then it's also overwritten. 
I guess it was part of some sort of optimization or buffer re-use before.
I had a commit that removed passing this useless `buf` around but de-scoped it from the PR.

sjanc commented:
#AutoPTS run zephyr L2CAP

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55519#issuecomment-1522863226 after 08:48:23.

codecoup-tester commented:
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/55519#issuecomment-1522863226 after 08:53:49.

codecoup-tester commented:
AutoPTS Bot results:
Failed tests:
L2CAP   L2CAP/ECFC/BV-43-C              FAIL - ERRATA TSE18351
L2CAP   L2CAP/ECFC/BV-44-C              INCONC - ERRATA TSE18351
L2CAP   L2CAP/TIM/BV-03-C               INCONC - ERRATA TSE22286

alwa-nordic commented:
DNM until we triage the failing PTS tests.
